[pytorch] Elide more Thrift Tensor send copies.

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31998 [pytorch] Elide more Thrift Tensor send copies.**

Note: This is a supporting 3-line change in Pickler to support a 170-line
change in Thrift serialization to elide copies in some cases.

Facebook:
Please see D19306682 for the bulk of the change.

Differential Revision: [D19306682](https://our.internmc.facebook.com/intern/diff/D19306682/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19306682/)!

Commit:
[pytorch] Elide more Thrift Tensor send copies.

This change builds on recent torch::from_blob() changes to avoid Tensor
copies on send in more cases.

Particularly, this change adds an enabled option to assume if the Tensor
Storage's DataPtr has a non-trivial deleter, then the Tensor does in fact
manage the underlying memory. And hence we can reference the Tensor's Storage
via an IOBuf that is referenced while sending, saving a Tensor copy.

We add appropriate test cases, particularly re: torch::from_blob() which
would have been problematic would recent changes.

Differential Revision: [D19306682](https://our.internmc.facebook.com/intern/diff/D19306682/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19306682/)!

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 4f57e98a:


**None of the build failures appear to be your fault**.

* **1/2** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=0dbd5c0bfe9c4af1ecb0fccc669277d6eafd0d2b&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 0dbd5c0b since Jan 08
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **1/2** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=4f57e98ae929717062314ab5ecacba65efc671ba).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4185933) caffe2_onnx_py2_gcc5_ubuntu16_04_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=82693798) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=82693798)) :snowflake:
<details>
<summary>
<code>Jan 09 19:24:57 E       RuntimeError: required keyword attribute 'Y_scale' is undefined</code>
</summary>

```
Jan 09 19:24:57 k = 'Y_scale' 
Jan 09 19:24:57  
Jan 09 19:24:57     def _node_getitem(self, k): 
Jan 09 19:24:57         r""" 
Jan 09 19:24:57         Accessor for attributes of a node which is polymorphic over 
Jan 09 19:24:57         return type. 
Jan 09 19:24:57      
Jan 09 19:24:57         NB: This is monkey-patched onto Node. 
Jan 09 19:24:57         """ 
Jan 09 19:24:57 >       sel = self.kindOf(k) 
Jan 09 19:24:57 E       RuntimeError: required keyword attribute 'Y_scale' is undefined 
Jan 09 19:24:57  
Jan 09 19:24:57 ../.local/lib/python2.7/site-packages/torch/onnx/utils.py:836: RuntimeError 
Jan 09 19:24:57 =============================== warnings summary =============================== 
Jan 09 19:24:57 /usr/local/lib/python2.7/dist-packages/scipy/_lib/_numpy_compat.py:10 
Jan 09 19:24:57   /usr/local/lib/python2.7/dist-packages/scipy/_lib/_numpy_compat.py:10: DeprecationWarning: Importing from numpy.testing.nosetester is deprecated since 1.15.0, import from numpy.testing instead. 
Jan 09 19:24:57     from numpy.testing.nosetester import import_nose 
Jan 09 19:24:57  
Jan 09 19:24:57 /usr/local/lib/python2.7/dist-packages/scipy/stats/morestats.py:12 
Jan 09 19:24:57   /usr/local/lib/python2.7/dist-packages/scipy/stats/morestats.py:12: DeprecationWarning: Importing from numpy.testing.decorators is deprecated since numpy 1.15.0, import from numpy.testing instead. 
Jan 09 19:24:57     from numpy.testing.decorators import setastest 
```

</details>

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_macos_10_13_py3_test](https://dr.pytorch.org/build-details.html?build_id=82693800) from Jan 08 until Jan 09



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jjlilley commented:
Any thoughts?

Commit:
Update on "[pytorch] Elide more Thrift Tensor send copies."


Note: This is a supporting 3-line change in Pickler to support a 170-line
change in Thrift serialization to elide copies in some cases.

Please see D19306682 for the bulk of the change.

Differential Revision: [D19306682](https://our.internmc.facebook.com/intern/diff/D19306682/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19306682/)!

[ghstack-poisoned]

jjlilley commented:
Rebasing this, since it's been sitting for a few weeks.

In micro benchmarks, this reduces user-space cpu by ~5% in SSL, and closer to ~20% for non-SSL, for 1M tensor echos.

Commit:
rebase on "[pytorch] Elide more Thrift Tensor send copies."


Note: This is a supporting 3-line change in Pickler to support a 170-line
change in Thrift serialization to elide copies in some cases.

Please see D19306682 for the bulk of the change.

Differential Revision: [D19306682](https://our.internmc.facebook.com/intern/diff/D19306682/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19306682/)!

[ghstack-poisoned]
